### ISSUE TITLE
Access modifiers limit subclassing

### ISSUE THREAD

**Issue by nmccann** [raulriera/TextFieldEffects#23]
Hi,

I'm using Cocoapods to add this project to my own, but that's causing issues when I want to make a custom subclass of TextFieldEffects since a few of the functions are declared as internal rather than public - all of the TextFieldsEffectsProtocol methods for example. I was able to bypass this by overriding the functions that were public (ex. drawRect, prepareForInterfaceBuilder) but that isn't ideal.

I was wondering if this limitation was intentional?

Thank you for the wonderful library!

**Comment by raulriera** [comment]
Hi, that was intentional, but I guess I could expose more API for them. Thanks for using it :)

**Comment by raulriera** [comment]
Could you point out some functions you would like to be public?

Thanks :)

**Comment by nmccann** [comment]
I think making all of the ones that throw `fatalError("\(__FUNCTION__) must be overridden")` should be sufficient.

At that point you could either make the `TextFieldsEffectsProtocol` public, or omit it altogether. It seems redundant since all of the default effects inherit from `TextFieldEffects`, and subclasses in other modules can't access the protocol. I can't think of a use case where you would conform to that protocol without inheriting from `TextFieldEffects`.

`placeholderLabel` also seems out of place - it's not used in `TextFieldEffects` and it's up to the subclasses to add it to their view hierarchy. I don't think it makes sense to define it in the base class.

**Comment by raulriera** [comment]
True, the `TextFieldsEffectsProtocol` was meant to be the interface for every "new" effect added to the library, but at the end everything turned into a subclass.. (this could be revisited in the Swift 2.0 version when Xcode 7 comes out)

Thanks again, I will improve this a lot for the Swift 2.0 release

**Comment by raulriera** [comment]
Removed the protocol in `master`, but I am a bit hesitant about opening the methods as public. It doesn't really make sense for people to have direct access to things like 

```swift
textFieldInstance.drawViewsInRect.updateViewsForBoundsChange(bounds)
```

Subclassing is something useful, but I don't see how those methods would help 😔 they will only make the "API" uglier. If you could share what functionality you want to add maybe I can come up with new methods or properties for it.

Thanks

**Comment by nmccann** [comment]
I understand your hesitation about exposing something like `updateViewsForBoundsChange` - it's a shame Swift doesn't have a `protected` access modifier.

The existing implementation of TextFieldEffects really seems like the classes within the library are special, and if you want to make changes to them (or make your own subclass), you have to get included in the library. Which might be intentional. The `TextFieldEffects` class isn't *really* public even though it's declared as such (probably because Swift doesn't allow subclasses to have greater accessibility than their superclass), because you can't override the functions that the `TextFieldEffects` class says must be overridden with `fatalErrors`.

If I didn't have access to the source `TextFieldEffects`, I wouldn't (easily) know that overriding `drawRect` and `prepareForInterfaceBuilder` would avoid the fatal calls to the internal function `drawViewsForRect`. Anyone looking to build on your library would be unable to do so without forking (or checking out) the library and developing in it. I prefer to develop within my own project, then extract the code into a library if it seems like it would be useful for others or if it's polished enough.

Now, the `TextFieldEffects` class isn't particularly complex on it's own so it might make more sense to reimplement it's logic within my own class rather than subclassing. But subclassing allows me to benefit from future improvements, bug fixes etc.

So to summarize: the `TextFieldEffects` class is public, but not obviously subclass-able, which is counter-intuitive.

**Comment by raulriera** [comment]
Gotcha, and you are correct... I should stop policing over the user and allow them to do what they want in order to built open this, it's not like iOS is preventing me from calling `drawRect:` directly when I subclassed UITextField to do this. I have committed 594f074d2 to make them all public, I am going to test it for a while and then make a release.

Thanks for the help

**Comment by nmccann** [comment]
Thanks @raulriera!

**Comment by raulriera** [comment]
Fixed in the 0.4.2 release
